### PR TITLE
clobber ~/.docker with a tmpfs when running planter

### PR DESCRIPTION
**Pull request by BenTheElder**
fixes https://github.com/kubernetes/test-infra/issues/5736

Solution also from https://github.com/kubernetes/test-infra/issues/5736 by ash2k

/area planter

### CODE REVIEW

**Comment by BenTheElder** [comment]
/assign @ixdy 
cc @ash2k 
supersedes https://github.com/kubernetes/test-infra/pull/5766

File: planter/planter.sh [hunk 1]
```diff
@@ -24,7 +24,7 @@ IMAGE="${IMAGE_NAME}:${TAG}"
 # run our docker image as the host user with bazel cache and current repo dir
 REPO=$(git rev-parse --show-toplevel 2>/dev/null || true)
 REPO=${REPO:-${PWD}}
-VOLUMES="-v ${REPO}:${REPO} -v ${HOME}:${HOME} --tmpfs /tmp:exec,mode=777"
+VOLUMES="-v ${REPO}:${REPO} -v ${HOME}:${HOME} --mount type=tmpfs,destination=${HOME}/.docker --tmpfs /tmp:exec,mode=777"
```
Comment (ixdy) [inline planter/planter.sh]: this change probably wants a comment explaining it.

also what's the difference between `--mount type=tmpfs` and `--tmpfs`?

Comment (BenTheElder) [reply planter/planter.sh]: SGTM

`--mount` is what docker's docs would prefer you to use, but `--tmpfs` actually allows you to set `:exec` 🤷‍♂️

Comment (BenTheElder) [reply planter/planter.sh]: I rewrote all of the comments, they were a bit stale and it seemed worth more generally explaining why we do things for anyone curious.

File: planter/planter.sh [hunk 2]
```diff
@@ -13,21 +13,44 @@
...
-# run our docker image as the host user with bazel cache and current repo dir
+# We want to mount our bazel workspace and the bazel cache
+# - WORKSPACE is assumed to be in your current git repo, or alternatively $PWD
 REPO=$(git rev-parse --show-toplevel 2>/dev/null || true)
 REPO=${REPO:-${PWD}}
-VOLUMES="-v ${REPO}:${REPO} -v ${HOME}:${HOME} --tmpfs /tmp:exec,mode=777"
+# - the bazel cache is in $HOME/.cache/bazel
+# - we could just mount $HOME/.cache/bazel, but unfortunately some rules
+# are still not terribly hermetic and actually need $HOME (rules_node)
+# - $HOME/.docker shouldn't be mounted so we clobber it with a tmpfs (issues/5736)
```
Comment (ixdy) [inline planter/planter.sh]: what uses `$HOME/.docker`?

Comment (BenTheElder) [reply planter/planter.sh]: you can have a `$HOME/.docker/config.json` for logging in to docker hub, which on mac will actually store credentials in your keychain which the container cannot access, this breaks rules_docker (see https://github.com/kubernetes/test-infra/issues/5736).
 
https://docs.docker.com/engine/reference/commandline/login/#extended-description

Comment (ixdy) [reply planter/planter.sh]: can you update the comment to note this?

Comment (BenTheElder) [reply planter/planter.sh]: done, also spaced out the bullets to hopefully make this more readable

**Pull request merged**